pkg/shell/indexes.js

```diff
@@ -332,13 +438,89 @@ function MachinesIndex(index_options, machines, loader, mdialogs) {
             return listItem;
         }
 
-        var menu = compiled.ordered("menu").map(links);
+        function keyword_relevance(current_best, item) {
+            const native = item.native || false;
+            const coef = item.coef || 0;
+            let score;
+            let _m = "";
+            let best = { score: -1 };
+            item.matches.forEach(m => {
+                if (native)
+                    _m = _(m);
+                score = -1;
+                // Best score when starts in native language
+                if (native && _m.indexOf(term) == 0)
+                    score = 4 + coef;
+                // Second best score when starts in English
+                else if (m.indexOf(term) == 0)
+                    score = 3 + coef;
+                // Substring consider only when at least 3 letters were used
+                else if (term.length >= 3) {
+                    if (native && _m.indexOf(term) >= 0)
+                        score = 2 + coef;
+                    else if (m.indexOf(term) >= 0)
+                        score = 1 + coef;
+                }
+                if (score > best.score) {
+                    best = { keyword: m, score: score };
+                }
+            });
+            if (best.score > current_best.score) {
+                current_best = { keyword: best.keyword, score: best.score, goto: item.goto || null };
+            }
+            return current_best;
+        }
+
+        function keyword_filter(item) {
+            if (!term)
+                return true;
+            const best_keyword = Object.values(item.keywords).reduce(keyword_relevance, { score:-1 });
+            if (best_keyword.score > -1) {
+                item.keyword = best_keyword;
+                return true;
+            }
+            return false;
+        }
+
+        var menu = compiled.ordered("menu")
+                .filter(keyword_filter)
+                .sort((a, b) => { return a.keyword.score - b.keyword.score })
```

Comment:
Please avoid the `{ return }` wrapper, the difference is a single simple expression.